### ISSUE TITLE
Shared `include`

### ISSUE THREAD

**Issue by AlexWayfer** [dazuma/toys#8]
Hello!

We're using `exec_tool` in many tools, and it'd be nice to write `include :exec, exit_on_nonzero_status: true` only once, somewhere in `.preload/` or `.toys/.toys.rb`, for example.

**Comment by dazuma** [comment]
Multiple people have asked for this. I'll try to prioritize it for 0.10.

Middleware is the current mechanism for affecting a whole set of tools. (i.e. it's how all tools get online help, and how all tools get `--verbose` and `--quiet` flags.) However, currently middleware is always static, defined globally on the CLI object.

The plan is to allow tools to affect the middleware stack of their subtool tree, and then include some syntactic sugar in the DSL that adds a middleware appending stuff to the end of every subtool. Usage might look something like this:

```ruby
# .toys.rb

# Apply this block to the end of every tool under the current tool
update_subtools do
  # Check to make sure the specific tool didn't already include :exec
  # before doing it here.
  include :exec, exit_on_nonzero_status: true unless include? :exec
end

tool :foo do
  def run
    sh "echo hello"
  end
  # This will have the above block added automatically
end
```

**Comment by AlexWayfer** [comment]
I'm not sure about `update_subtools`.

If sub-tools already inherit methods and other staff… maybe it should inherit included middlewares / modules?

**Comment by dazuma** [comment]
Subtools don't inherit methods, or anything else. They're namespaces only.

**Comment by dazuma** [comment]
Actually, no I'm wrong. They do inherit definitions of *things that can be looked up* like mixins and acceptors. But things that actually define functionality, like methods or flags, are explicitly kept separate.

**Comment by dazuma** [comment]
Closing this as fixed, since the code is in master. I expect to release this in version 0.10.0 shortly.

**Comment by AlexWayfer** [comment]
I have `include :exec, exit_on_nonzero_status: true` in the `./toys/.toys.rb`, and I'm getting:

```
NoMethodError: undefined method `exec_tool' for #<#<Class:0x0000558cb564b328>:0x0000558cb57395c8>
    in config file: /home/alex/project/.toys/database/migrations/run.rb:9
    while executing tool: "database migrations run"
    with arguments: []
```

**Comment by dazuma** [comment]
Would you be willing to share a gist (or email me if you want to keep it private) of your `.toys.rb` file?

(Note: the actual final implementation names the block `subtool_apply` rather than `update_subtools`.)

**Comment by dazuma** [comment]
The other thing I can think of is, it works only if the `subtool_apply` is declared in the same ultimate `.toys` directory where the tool is defined. So it should work if it's in `/home/alex/project/.toys/.toys.rb` but won't if it's in `/home/alex/.toys/.toys.rb`. (This is intentional. It can lead to a lot of confusion if something you do in one directory can affect tools globally in other directories. The same scoping rule applies when doing things like defining named mixins or acceptors.)

**Comment by AlexWayfer** [comment]
Sorry, with `subtool_apply` it works. I guessed there is automatic inheritance.

**Comment by dazuma** [comment]
The rationale behind all this, is for tools to appear to behave as much like classes as possible.  Most things don't "inherit" into nested modules, and the behavior of nested tools is analogous.

For example, if I define a method `Foo#helper`, it does not automatically get added to class `Foo::Bar`. Similarly, if class `Foo` does an `include MyModule`, that does not affect class `Foo::Bar`. Nested tools behave the same way. Defining a method in tool `foo` does not add it to tool `foo bar`, and including a mixin in tool `foo` also does not automatically include it in `foo bar`. (So a special mechanism `subtool_apply` is needed to apply a mixin to multiple tools, just like you'd have to write some special code in order to include a module in multiple nested classes.)

That said, things that are looked up by name from tools (such as named acceptors) *are* visible from nested tools. This is because the behavior is analogous to constants in Ruby: if I define a *constant* `Foo::MY_CONST`, that *is* accessible from `Foo::Bar` because of constant lookup rules.

So that's my reasoning for the current behavior. The principle is to remain close to the behavior of existing Ruby constructs. I'm open to being convinced we should be doing it differently, but currently I think adhering to expectations of Ruby developers is a pretty strong argument.

**Comment by AlexWayfer** [comment]
@dazuma OK, thank you, this example is clear enough. But… I can't construct an analogue for `Bar < Foo`, which will has `#helper` automatically. We have `include`, we have namespaces, but what about inheritance?

Because of `Foo` is a tool itself, so it's like class, not module-for-namespace, I guess the following analogue code is appropriate:


```ruby
class Foo < Toys::Tool
  def run
    # do something
  end

  private

  def helper
    # something common
  end

  # Here is you're expecting `< Toys::Tool` again,
  # but why not `< self`?
  class Bar < self
    def run
      # do another thing
      # `helper` is available
    end
  end
end
```

Yes, `run` will be inherited too, but it's not hard to check if it defined explicitly or inherited with flags for Ruby methods like `instance_method_defined?`.

**Comment by dazuma** [comment]
Yeah, Toys doesn't have an analogue of subclasses/inheritance, because in the context of command-line tools, I thought it seemed better to model sharing via composition (e.g. mixins) rather than inheritance. For some of the reasons you mentioned: would it inherit the `run` method, the flags, the description, etc? which all seems not very useful. We could selectively omit certain things, but then how do we choose what to omit and what to keep, and is there more potential for confusion there?

I get that you *could* say `class Bar < self`. But how often do you do that in Ruby? Maybe more to the point, is it often enough that you would do it *by default* without any extra syntax:

```ruby
class Foo
  class Bar  # does it make sense to be "< self" by default?
  # ...
```

So I would hesitate for tools to *always* inherit. So then, perhaps we could provide some extra syntax (i.e. similar to adding `< self` for a nested class) like this:

```ruby
tool "foo" do
  tool "bar", inherit: true do
  # ...
```

but then, is that really significantly better than:

```ruby
tool "foo" do
  tool "bar" do
    include "shared_stuff"
    # ...
```

**Comment by AlexWayfer** [comment]
> For some of the reasons you mentioned: would it inherit the run method, the flags, the description, etc? which all seems not very useful. We could selectively omit certain things, but then how do we choose what to omit and what to keep, and is there more potential for confusion there?

Fair enough.

> I get that you could say class Bar < self. But how often do you do that in Ruby? Inherit from your namespace? Maybe more to the point, is it often enough that you would do it by default without any extra syntax

To be honest, I have [Flame web-framework](https://github.com/AlexWayfer/flame), and there are controllers with it looks like:

```ruby
module MyApplication
  class Controller < Flame::Controller
    private

    def some_common_stuff
    end
  end

  class IndexController < Controller # MyApplication::Controller
    def index; end
  end

  module NestedNamespace
    class Controller < MyApplication::Controller
      private

      def another_common_stuff; end
    end

    class FooController < Controller # NestedNamespace::Controller
      def page; end
    end
  end
end
```

So, there are private methods as helpers, and public methods as actions. And only defined public methods. So, if we have some module with common actions (public methods), we should do `include with_actions CommonAction`, which will re-define public methods from `CommonAction` in a class in which it'll be included.

I don't mean that this is ultimate super wonderful system, etc., but you can just consider it. It works in pretty large commercial code bases.

> `tool "bar", inherit: "base" do`

Please, think about tool-as-file system, where is name of tool equals to name of file.

**Comment by AlexWayfer** [comment]
So, what about `tool "bar", inherit: true`? It could simplify my life. Should I create a new issue?

**Comment by AlexWayfer** [comment]
Also I have troubles with passing template arguments to nested tools.

We have [Template classes](https://dazuma.github.io/toys/gems/toys/v0.10.1/file.guide.html#Template_classes), let's imagine a basic one, parent, without `on_expand`. 

And we have some templates inherited from a basic one and with `on_expand`.

How to pass an argument from `Basic#initialize` to `on_expand do; tool :foo do;`?

**Comment by dazuma** [comment]
Hi @AlexWayfer , sorry I missed your last couple of comments. Can you open new issues for these? (Feel free to link back to this issue for context, if applicable.) When someone comments on a closed issue, I often miss it or forget about it because closed issues don't show up on the issue list in GitHub's UI by default.

**Comment by AlexWayfer** [comment]
> When someone comments on a closed issue, I often miss it or forget about it because closed issues don't show up on the issue list in GitHub's UI by default.

OK, thank you, I didn't know that because I'm using notifications from emails.

> Can you open new issues for these? (Feel free to link back to this issue for context, if applicable.)

Yes, I can.

**Comment by AlexWayfer** [comment]
Also, I've found and discovered some answers for my questions:


> How to pass an argument from `Basic#initialize` to `on_expand do; tool :foo do;`?

Just use `on_expand do |template|;`, then `to_run do` instead of `def run` for local variables access, and then `template.foo_variable`.
